router/src/main/kotlin/io/moia/router/PermissionHandler.kt

```diff
@@ -0,0 +1,68 @@
+package io.moia.router
+
+import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+import java.util.Base64
+
+interface PermissionHandler {
+
+    fun hasAnyRequiredPermission(requiredPermissions: Set<String>): Boolean
+}
+
+class NoOpPermissionHandler : PermissionHandler {
+    override fun hasAnyRequiredPermission(requiredPermissions: Set<String>) = true
+}
+
+open class JwtAccessor(
+    private val request: APIGatewayProxyRequestEvent,
+    private val authorizationHeaderName: String = "authorization"
+) {
+
+    private val objectMapper = jacksonObjectMapper()
+
+    fun extractJwtToken(): String? =
+        // support "Bearer <token>" as well as "<token>"
+        request.getHeaderCaseInsensitive(authorizationHeaderName)?.split(" ")?.toList()?.last()
+
+    fun extractJwtClaims() =
+        extractJwtToken()
+            ?.let { token -> token.split("\\.".toRegex()).dropLastWhile { it.isEmpty() } }
+            ?.takeIf { it.size == 3 }
+            ?.let { it[1] }
+            ?.let { jwtPayload ->
+                try {
+                    String(Base64.getDecoder().decode(jwtPayload))
+                } catch (e: Exception) {
+                    return null
+                }
+            }
+            ?.let { objectMapper.readValue<Map<String, Any>>(it) }
+}
+open class JwtPermissionHandler(
+    val accessor: JwtAccessor,
+    val permissionsClaim: String = "scope",
+    val permissionSeparator: String = " "
+) : PermissionHandler {
+
+    constructor(
+        request: APIGatewayProxyRequestEvent,
+        permissionsClaim: String = "scope",
+        permissionSeparator: String = " "
+    ) : this(JwtAccessor(request), permissionsClaim, permissionSeparator)
+
+    override fun hasAnyRequiredPermission(requiredPermissions: Set<String>): Boolean =
+        extractPermissions().any { requiredPermissions.contains(it) }
```

Comment:
This implementation only cares about the contents of the `permissions` claim. So for all the other verification of the basic claims like `exp,aud,iss` and the signature this implementation relies on the API Gateway to do it's job.

I think this won't be a problem at all, as we're calling our lambdas from the API Gatrway only (do we?) but I just wanted to point out that this implementation is only secure for calls behind the API Gateway.

Reply:
Yes good point. This is exactly the assumption I made.

Reply:
Good point. Maybe that is something that should be added to the README.

Reply:
I added a hint to the README